### PR TITLE
Refactor fetchEntries, date checking code, new Exception classes, more unit tests

### PR DESCRIPTION
**Pull request by jameswenzel**
- refactor fetchEntries with more digestible helper methods
- use a single `Requests.Session` object for all outgoing requests. `Session`s store cookies between requests and look less like bots to websites. From personal experience, it seems like Billboard has imposed some rate limiting, and this may help with that
- date checking code and more unit tests (#38) c/o @johannfrias (#40)
- Introduce a generic `BillboardException` that all other exceptions are subclasses of. Introduce `BillboardContentException` and `BillboardEmptyResponseException` (which inherits from the former). `BillboardNotFoundException` now inherits from `BillboardContentException`. This allows for catching of generic `BillboardException`s, or separate handling of `BillboardParseException`s (problems parsing information) and `BillboardContentException`s (problems retrieving information in the first place).

### CODE REVIEW

**Comment by jameswenzel** [comment]
I believe the CI Failure is due to billboard rate-limiting. It might pass if run again.

**Comment by guoguo12** [comment]
Pretty sure there's an actual bug:

```
>>> billboard.ChartData('latin-pop-songs', date='1994-10-15')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/allen/billboard-charts/billboard.py", line 170, in __init__
    self.fetchEntries()
  File "/home/allen/billboard-charts/billboard.py", line 337, in fetchEntries
    raise BillboardEmptyResponseException(message)
billboard.BillboardEmptyResponseException: Billboard returned a blank page. (Has this chart been released yet?)
```

This doesn't seem quite right. Here's the chart: https://www.billboard.com/charts/latin-pop-songs/1994-10-15.

**Comment by guoguo12** [comment]
Sorry it's been so long, by the way.

File: billboard.py [hunk 1]
```diff
@@ -140,8 +154,12 @@ def __init__(self, name, date=None, fetch=True, timeout=25):
         """
         self.name = name
 
-        if date is not None and not re.match('\d{4}-\d{2}-\d{2}', str(date)):
-            raise ValueError('Date argument is not in YYYY-MM-DD format')
+        if date is not None:
+            try:
+                datetime.datetime(*(int(x) for x in str(date).split('-')))
+            except ValueError as e:
+                raise ValueError('Date argument is invalid. ' + str(e))
```
Comment (guoguo12) [inline billboard.py]: The old error message was more helpful.

File: billboard.py [hunk 2]
```diff
@@ -10,16 +10,18 @@
 
 """billboard.py: Unofficial Python API for accessing music charts from Billboard.com."""
 
-__author__     = "Allen Guo"
-__license__    = "MIT"
+__author__ = "Allen Guo"
```
[the remark this answers is not in the document]
Comment (jameswenzel) [reply billboard.py]: Yeah, must have been picked up by my linter. I really need to dig into those settings...

File: billboard.py (hunk 1 above)
Comment (jameswenzel) [reply billboard.py]: Agreed

**Comment by guoguo12** [comment]
> Can you see entries 2-onward?

Nope, I can only see the top song.

Not really sure what to make of that.

But I think the right behavior is to just have a chart with one entry. That seems like the safest option. From a user's perspective, it would be confusing if the library disagreed with the website for no obvious reason.

**Comment by guoguo12** [comment]
Okay, agreed. A chart with only a single entry should never be considered valid, so raising is the right move.

I reran the test suite and it's still failing. Can you remove (or ideally replace) the offending test? It looks like the Latin pop songs chart is broken on Billboard's end.

File: billboard.py [hunk 3]
```diff
@@ -236,69 +244,108 @@ def fetchEntries(self):
...
+        return ChartEntry(topTitle, topArtist, topPeakPos,
+                          topLastPos, topWeeks, 1, topIsNew)
 
-        for entrySoup in soup.select(_ENTRY_LIST_SELECTOR):
-            try:
-                title = entrySoup[_ENTRY_TITLE_ATTR].strip()
-            except:
-                message = "Failed to parse title"
-                raise BillboardParseException(message)
+    def getPositionRowValue(self, entrySoup, rowName):
```
Comment (guoguo12) [inline billboard.py]: This should be `_getPositionRowValue` with an underscore, right?

**Pull request closed without merging**